src/quantization/scalar_quantization_trainer.cpp

```diff
@@ -0,0 +1,100 @@
+
+// Copyright 2024-present the vsag project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "scalar_quantization_trainer.h"
+
+#include <random>
+#include <unordered_set>
+
+#include "simd/normalize.h"
+
+namespace vsag {
+
+ScalarQuantizationTrainer::ScalarQuantizationTrainer(int32_t dim, int bits)
+    : dim_(dim), bits_(bits) {
+}
+
+void
+ScalarQuantizationTrainer::Train(const float* data,
+                                 uint64_t count,
+                                 float* upper_bound,
+                                 float* lower_bound,
+                                 bool need_normalize,
+                                 ScalarQuantizationTrainer::SQTrainMode mode) {
+    std::vector<float> sample_datas;
+    auto sample_count = this->sample_train_data(data, count, sample_datas, need_normalize);
+    if (mode == CLASSIC) {
+        this->classic_train(sample_datas.data(), sample_count, upper_bound, lower_bound);
+    }
+}
+
+void
+ScalarQuantizationTrainer::TrainUniform(const float* data,
+                                        uint64_t count,
+                                        float& upper_bound,
+                                        float& lower_bound,
+                                        bool need_normalize,
+                                        ScalarQuantizationTrainer::SQTrainMode mode) {
+    std::vector<float> sample_datas;
+    auto sample_count = this->sample_train_data(data, count, sample_datas, need_normalize);
+    std::vector<float> upper(dim_);
+    std::vector<float> lower(dim_);
+    if (mode == CLASSIC) {
+        this->classic_train(sample_datas.data(), sample_count, upper.data(), lower.data());
+        upper_bound = *std::max_element(upper.begin(), upper.end());
+        lower_bound = *std::min_element(lower.begin(), lower.end());
+    }
+}
+
+void
+ScalarQuantizationTrainer::classic_train(const float* data,
+                                         uint64_t count,
+                                         float* upper_bound,
+                                         float* lower_bound) {
+    for (uint64_t i = 0; i < dim_; ++i) {
+        upper_bound[i] = std::numeric_limits<float>::lowest();
+        lower_bound[i] = std::numeric_limits<float>::max();
+        for (uint64_t j = 0; j < count; ++j) {
+            auto value = data[j * dim_ + i];
+            upper_bound[i] = std::max(upper_bound[i], value);
```

Comment:
minor: Here change to check and update, is it possible to reduce unnecessary memory write ?

Reply:
I think compiler may optimize it, the upper_bound[i] may in register